design/design.md

```diff
@@ -0,0 +1,232 @@
+# Design for Streamly Coreutils package
+
+* cp
+
+  ```
+      data CpOptions = CpOptions {
+         attributesOnly :: Bool,
+         interactive :: Bool,
+         parents :: Bool,
+         symbolicLink :: Bool,
+         noTargetDir :: Bool,
+         verbose :: Bool,
+         noTargetDir :: Bool
+         ...
+      }
+
+      cpFile :: CpOptions -> SomeBase File -> SomeBase Dir -> SomeBase File -> IO ()
```

Comment:
Is this the correct signature?

Reply:
I thought of this signature to combine the `File -> File` and `File -> Dir` cases.

Reply:
How would something like this work: (assume types paths)
`cpFile "/test.txt" "/new_dir/" "/some_other_dir/test2.txt"`
Would'nt you agree this to be a better signature: 
`cpFile :: CpOptions -> SomeBase File -> SomeBase Dir -> String -> IO ()`
A better option would be to ignore the destination file name. Start by creating simple functions.

Reply:
Oh okay I meant the same thing. :+1: 
First argument : file to be copied, second : destination directory, third : the new file name, which should be a string as you pointed out.

Reply:
I think we should keep the destination file name.

Reply:
You can, just not in the basic functions. `cpFileWithRename :: Path File -> Path File` makes more sense if you want to use a destination file name. I'm fine with either for now.

Reply:
So should we go ahead with

`cpFile :: CpOptions -> Path File -> Path Dir -> String -> IO ()`

The String argument is for the new file. The user can specify the same filename as of the source file if they do not want a rename.